graviton/invariant.py

```diff
@@ -0,0 +1,155 @@
+from inspect import signature
+from typing import Any, Callable, Dict, List, Tuple, Union
+
+from graviton.blackbox import (
+    DryRunInspector,
+    DryRunProperty,
+    ExecutionMode,
+    mode_has_property,
+)
+
+
+class Invariant:
+    """Enable asserting invariants on a sequence of dry run executions"""
+
+    def __init__(
+        self,
+        predicate: Union[Dict[Tuple, Union[str, int]], Callable],
+        enforce: bool = False,
+        name: str = None,
+    ):
+        self.definition = predicate
+        self.predicate, self._expected = self.prepare_predicate(predicate)
+        self.enforce = enforce
+        self.name = name
+
+    def __repr__(self):
+        return f"Invariant({self.definition})"[:100]
+
+    def __call__(self, args: list, actual: Union[str, int]) -> Tuple[bool, str]:
+        invariant = self.predicate(args, actual)
+        msg = ""
+        if not invariant:
+            msg = f"Invariant for '{self.name}' failed for for args {args}: actual is [{actual}] BUT expected [{self.expected(args)}]"
+            if self.enforce:
+                assert invariant, msg
+
+        return invariant, msg
+
+    def expected(self, args: list) -> Union[str, int]:
+        return self._expected(args)
+
+    def validates(
+        self,
+        property: DryRunProperty,
```

Comment:
@tzaffi Minor:  IDE hints that `property` shadows a built-in name.  Unless you feel strongly, it seems prudent to rename to avoid unintended side-effects.  

If you prefer to keep as is, then let's resolve the thread.

Reply:
Another good catch!